Allow element_proc to be passed via options or html_options

rubiii opened the pull request:
@wooly @Spone Here's a first draft of the proposed change.

I moved the code to extract the `element_proc` into a new method and it now raises if it receives two `element_proc` options. This could be extracted somewhere, but I'm not sure if there's already a good place for it. I thought about introducing a `CollectionFieldComponent`, but it feels too big of a change.

Also I'm not sure why `set_html_options!` is a bang method, but I followed the naming convention? and named the new method accordingly.

Let me know what you think, what needs to be changed, ...
and enjoy your holidays!

rubiii committed:
Allow element_proc to be passed via options or html_options

Spone reviewed (approved):
Looking good!

rubiii commented:
FYI: Rubocop complained about `too many memoized helpers` (the limit is at 7 per spec), so I changed the main component let to a subject, because it's the subject under test and used before blocks instead of scoped memoized helpers for the new specs. Code was rearranged by Rubocop as well.

Spone commented:
Another way to keep the code DRY would be to move the `element_proc` logic to a concern that would be included in both components.

rubiii commented:
Sure. If I'm not missing anything, it seems like this would be the first concern for components then? Do you have any preferences about where to place and how to name it?

`lib/view_components/form/element_proc_concern.rb` maybe?
Which would contain the `set_element_proc!` method, the `element_proc` attr_reader and an initialize method calling super and `set_element_proc!`?

Spone commented:
Yes it would be the first concern.

In my Rails apps using ViewComponent, I usually place the component concerns at `app/components/concerns`, similar to the controllers or models concerns.

So maybe we could have `app/components/concerns/view_component/form/element_proc.rb` that defines a `ViewComponent::Form::ElementProc` module.

> Which would contain the `set_element_proc!` method, the `element_proc` attr_reader and an initialize method calling super and `set_element_proc!`?

Yes, that sounds good.

rubiii committed:
Extract ElementProc concern

rubiii committed:
Fix autoload error

Zeitwerk::Error:
  loader

<some Zeitwerk::Loader> wants to manage directory
app/components/concerns, which is already managed by
<other Zeitwerk::Loader>

rubiii commented:
@Spone I pushed the proposed change. Adding the concern seems to require a change to autoload paths for Zeitwerk not to blow up. Otherwise it seems to work fine. Let me know if that would work for you.
